File: src/core/img_proc.go
package main

import (
	"errors"
	"fmt"
	"image"

	"gocv.io/x/gocv"
)

// func main() {
// 	img := gocv.IMRead("data/test.png", gocv.IMReadColor)
// 	fmt.Println(img.Size())
// 	blur, err := preprocess(img)
// 	if err != nil {
// 		fmt.Println("error in preprocess")
// 		fmt.Println(err)
// 	}
// 	gocv.IMWrite("data/res.png", blur)
// 	flag, r, err := keyPointsColor(blur, 50)
// 	fmt.Println(flag, r)
// 	a, b, err := mainColor(blur)
// 	fmt.Println(a, b)

// }

func maxInSlice(arr []float64) float64 {
	var max float64 = -1.0
	for _, v := range arr {
		if v > max {
			max = v
		}
	}
	return max
}

func minInSlice(arr []float64) float64 {
	var min float64 = 256.0
	for _, v := range arr {
		if v < min {
			min = v
		}
	}
	return min
}

func sumInSlice(arr []float64) float64 {
	var sum float64 = 0.0
	for _, v := range arr {
		sum += v
	}
	return sum
}

// from https://github.com/hybridgroup/gocv/issues/833
func GetVecbAt(m gocv.Mat, row int, col int) []uint8 {
	ch := m.Channels()
	v := make([]uint8, ch)

	for c := 0; c < ch; c++ {
		v[c] = m.GetUCharAt(row, col*ch+c)
	}

	return v
}

func preprocess(img gocv.Mat) (gocv.Mat, error) {
	gray := gocv.NewMat()
	blur := gocv.NewMat()
	shape := img.Size()
	fmt.Println(shape)
	switch {
	case len(shape) == 3 && img.Size()[2] == 3:
		gocv.CvtColor(img, &gray, gocv.ColorBGRToGray)
	case len(shape) == 3 && img.Size()[2] == 4:
		gocv.CvtColor(img, &gray, gocv.ColorBGRAToGray)
	case len(img.Size()) == 2:
		gray = img
	default:
		return blur, errors.New("wrong input.")
	}
	gocv.CvtColor(img, &gray, gocv.ColorBGRToGray)
	gocv.GaussianBlur(gray, &blur, image.Point{3, 3}, 0.0, 0.0, gocv.BorderDefault)
	return blur, nil
}

func keyPointsColor(grayImg gocv.Mat, drift int8) (bool, float64, error) {
	h, w := grayImg.Size()[0], grayImg.Size()[1]
	var hBound int
	var wBound int
	if w <= 10 || h <= 10 {
		return false, 0, errors.New("image size ({}, {}) too small!")
	}
	if h < 100 {
		hBound = 3
	} else {
		hBound = 5
	}
	if w < 100 {
		wBound = 3
	} else {
		wBound = 5
	}
	topLeft := grayImg.Region(image.Rect(0, 0, wBound, hBound)) // left, top, right, bottom
	topRight := grayImg.Region(image.Rect(w-wBound, 0, w, hBound))
	bottomLeft := grayImg.Region(image.Rect(0, h-hBound, wBound, h))
	bottomRight := grayImg.Region(image.Rect(w-wBound, h-hBound, w, h))
	var keyPoints = [4]float64{
		topLeft.Mean().Val1,
		topRight.Mean().Val1,
		bottomLeft.Mean().Val1,
		bottomRight.Mean().Val1,
	}
	if maxInSlice(keyPoints[:])-minInSlice(keyPoints[:]) < float64(drift) {
		r := sumInSlice(keyPoints[:]) / 4
		return true, r, nil
	} else {
		return true, 0, nil
	}
}

// calculate main color at corners.
func mainColor(grayImg gocv.Mat) (uint8, float32, error) {
	const bins int = 256
	var colorHist [bins]float32
	var pixelNum float32 = 0
	var maxColor uint8 = 0
	var maxRatio float32 = 0

	mask := gocv.NewMat()
	ch := []int{0}
	src := []gocv.Mat{grayImg}
	hist := gocv.NewMat()
	size := []int{bins}
	ranges := []float64{0, 256}
	gocv.CalcHist(src, ch, mask, &hist, size, ranges, false)

	for i := 0; i < hist.Size()[1]; i++ { // row
		for j := 0; j < hist.Size()[0]; j++ { // col
			v := hist.GetFloatAt(i, j)
			pixelNum += v
			colorHist[j] = v
		}
	}
	for i := 0; i < bins; i++ {
		r := float32(colorHist[i]) / float32(pixelNum)
		if r > maxRatio {
			maxColor = uint8(i)
			maxRatio = r
		}
	}
	fmt.Println(pixelNum)

	return maxColor, maxRatio, nil
}

func generateMask(img gocv.Mat, th float32, drift int8) (gocv.Mat, error) {
	mask := gocv.NewMat()
	gray, err := preprocess(img)
	if err != nil {
		return mask, err
	}
	isSame, keyColor, err := keyPointsColor(gray, drift)
	maxColor, maxRatio, err := mainColor(gray)
	if isSame == false {
		fmt.Println("Warning: key points color not same!")
		return mask, errors.New("key points color not same!")
	}
	if uint8(keyColor) != maxColor {
		fmt.Println("Warning: key color != max color!")
	}
	if maxRatio > th {
		// TODO: finish this(marquee_todo)

	} else {

	}
	return mask, nil
}

func blend(img gocv.Mat, mask gocv.Mat) (gocv.Mat, error) {
	rgba := gocv.NewMat()
	img.CopyToWithMask(&rgba, mask)
	return rgba, nil
}

File: src/core/img_proc.py
import cv2
import numpy as np
import logging

log = logging.getLogger("bluefin-tuna")


def preprocess(img):
    if len(img.shape) == 3 and img.shape[2] == 3:
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    elif len(img.shape) == 3 and img.shape[2] == 4:
        img = img[:, :, 0:3]
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    elif len(img.shape) == 2:
        gray = img
    else:
        raise ValueError("wrong input image!")
    blur = cv2.GaussianBlur(gray, (3, 3), 0)
    return blur


def key_points_color(gray_img, drift=50):
    h, w = gray_img.shape[0], gray_img.shape[1]
    if w <= 10 or h <= 10:
        raise ValueError("image size ({}, {}) too small!".format(w, h))
    if h < 100:
        h_bound = 3
    else:
        h_bound = 5
    if w < 100:
        w_bound = 3
    else:
        w_bound = 5

    key_points = [np.mean(gray_img[0:h_bound, 0:w_bound]),
                  np.mean(gray_img[h - h_bound - 1:-1, 0:w_bound]),
                  np.mean(gray_img[0:h_bound, w - w_bound - 1:-1]),
                  np.mean(gray_img[h - h_bound - 1:-1, w - w_bound - 1:-1])]
    log.debug(key_points)
    if max(key_points) - min(key_points) < drift:
        return True, sum(key_points) / len(key_points)
    else:
        return False, 0


def main_color(gray_img):
    bins = 256
    h, w = gray_img.shape[0], gray_img.shape[1]
    pixel_num = h * w
    color_map = {}
    for i in range(bins):
        color_map[i] = len(np.where(gray_img == i)[0]) / pixel_num
    max_color = -1
    max_ratio = 0
    for k, v in color_map.items():
        if v > max_ratio:
            max_color = k
            max_ratio = v
    return max_color, max_ratio


def generate_mask(img, th=0.55, drift=20):
    gray = preprocess(img)
    mask = np.ones(gray.shape) * 255
    is_same, key_color = key_points_color(gray)
    max_color, max_ratio = main_color(gray)
    log.debug("max_ratio: {}".format(max_ratio))
    if not is_same:
        log.debug("Warning: key points color not same!")
        raise ValueError("key points color not same!")
    if key_color != max_color:
        log.debug("Warning: key color != max color!")
    if max_ratio > th:
        mask[((key_color - int(drift / 2)) <= gray) & (gray <= (key_color + int(drift / 2)))] = 0
        mask[((max_color - int(drift / 2)) <= gray) & (gray <= (max_color + int(drift / 2)))] = 0
    else:
        mask[gray == key_color] = 0
    return mask


def blend(img, mask):
    if img.shape[0:2] != mask.shape:
        raise ValueError("image shape != mask shape: image shape: {}, mask shape:{}".format(img.shape[0:2], mask.shape))
    rgba = np.zeros((mask.shape[0], mask.shape[1], 4), dtype=int)
    rgba[:, :, 0:3] = img[:, :, 0:3]
    rgba[:, :, 3] = mask
    return rgba